Click Tolerance for Markers

ChristopherChudzicki opened the pull request:
Implements a `clickTolerance` option for `Marker`s, as discussed in https://github.com/mapbox/mapbox-gl-js/issues/9636. Click tolerance is optionally set on a per-marker basis. If not set, the map's click tolerance is used.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog></changelog>`

#### Changelog Entry:

<!-- Unsure how to apply the changelog label? -->
<changelog label="feature">
Added `clickTolerance` as an option for `Marker`s. Pointer movement below `clickTolerance` does not fire `Marker` drag events.
</changelog>

ChristopherChudzicki commented:
@ryanhamley I believe that you described this issue as a bug in https://github.com/mapbox/mapbox-gl-js/issues/9624. It could also be considered a feature and a breaking change. (Would be non-breaking if the default tolerance was 0px instead of inherited from map). Anyway, I guess all bug fixes are breaking changes. I'll let you all decide what to do with it.

As also discussed in https://github.com/mapbox/mapbox-gl-js/issues/9624, I think there should be a separate issue / PR to suppress map click events when dragging markers. Currently, if pointer movement is ABOVE the marker click tolerance but BELOW the map click tolerance, then marker fires drag events AND map fires click events. That seems undesirable. (This really is a separate issue from this PR, and is currently present on master, where marker click tolerance is effectively 0px.)

Thanks for considering this PR, and for mapbox!

PS: The commit history in this PR looks a little out of order and I do not know why. It is in the correct order on my fork: https://github.com/ChristopherChudzicki/mapbox-gl-js/commits/marker-clicktol

andrewharvey commented on src/ui/marker.js:
```diff
@@ -36,6 +37,7 @@ type Options = {
  * @param {string} [options.color='#3FB1CE'] The color to use for the default marker if options.element is not provided. The default is light blue.
  * @param {number} [options.scale=1] The scale to use for the default marker if options.element is not provided. The default scale corresponds to a height of `41px` and a width of `27px`.
  * @param {boolean} [options.draggable=false] A boolean indicating whether or not a marker is able to be dragged to a new position on the map.
+ * @param {?number} [options.clickTolerance=null] A number specifying the click tolerance in pixels. For draggable markers, movement must exceed this tolerance to be treated as a drag. A `null` value inherits click tolerance from `Map`.
```
What is click tolerance here? Is it pixels away from the marker that are still counted as clicking the marker or is a small drag while holding the mouse down within the tolerance is still counted as a click and not a drag?

They are two different things and I can see users getting confused which one this affects.

andrewharvey replied:
I see from the changelog entre this would be the latter. I think we should make this clearer in the description to avoid people getting confused it's the former.

ChristopherChudzicki replied:
Yes, that is a good clarification. I do still think it should be called `clickTolerance` because that is the name used for the `Map` option. (The ambiguity you allude to doesn't really exist for the map, though.) 

I'll change the jsdoc description here to mirror the map description more closely:

> The max number of pixels a user can shift the mouse pointer during a click for it to be considered a valid click (as opposed to a mouse drag).

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -36,6 +37,7 @@ type Options = {
  * @param {string} [options.color='#3FB1CE'] The color to use for the default marker if options.element is not provided. The default is light blue.
  * @param {number} [options.scale=1] The scale to use for the default marker if options.element is not provided. The default scale corresponds to a height of `41px` and a width of `27px`.
  * @param {boolean} [options.draggable=false] A boolean indicating whether or not a marker is able to be dragged to a new position on the map.
+ * @param {?number} [options.clickTolerance=null] The max number of pixels a user can shift the mouse pointer during a click on the marker for it to be considered a valid click (as opposed to a marker drag).
```
I would **love** not to muck around with null/undefined types and I will plan to make this change. The reason I did not do it this way initially is that using `0` to mean "inherit from the map" means that the `v1.11.0` behavior cannot be reproduced.

That said, the `v1.11.0` behavior (where markers have 0 tolerance and map has 3px by default) seems undesirable to me, and people could always set the tolerance to be 0.00001.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -483,7 +490,49 @@ export default class Marker extends Evented {
+    _hasMovedBeyondClickTolerance(point: Point) {
+        const clickTolerance = this.getClickTolerance();
+        if (clickTolerance === null) return false;
+        return point.distSqr(this._pointerdownPos) >= clickTolerance ** 2;
```
> Computes the squared distance from this vector to v. If you are just comparing the distance with another distance, you should compare the distance squared instead as it is slightly more efficient to calculate.

That's from the [ThreeJS documetnation for distanceToSquared](https://threejs.org/docs/#api/en/math/Vector3.distanceToSquared) (for 3d webgl stuff). Obviously Mapbox isn't threejs, but when I saw that Mapbox `Point` objects have a `distSqr` method, I figured Mapbox encouraged similar squaring.

Even though the onMove handler is called frequently, this is very likely a premature optimization. I'm certainly happy to remove it.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -483,7 +490,49 @@ export default class Marker extends Evented {
+    }
+
     _onMove(e: MapMouseEvent | MapTouchEvent) {
+        if (!this._isDragging) {
+            this._isDragging = this._hasMovedBeyondClickTolerance(e.point);
+        }
+        if (!this._isDragging) return;
+
```
@ryanhamley I think the `_isDragging` boolean is necessary. What we want to check is
 - "has pointer movement ever exceeded the threshhold while pointer down" not,
 - "is current pointer beyond threshhold".
The latter condition results is jerky drag if the marker is dragged back to near where it started.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -58,8 +60,11 @@ export default class Marker extends Evented {
     _scale: number;
     _defaultMarker: boolean;
     _draggable: boolean;
+    _clickTolerance: ?number;
```
I belive this is resolved.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -86,6 +91,8 @@ export default class Marker extends Evented {
         this._color = options && options.color || '#3FB1CE';
         this._scale = options && options.scale || 1;
         this._draggable = options && options.draggable || false;
+        this._clickTolerance = options && options.clickTolerance || null;
```
I believe this is resolved.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -483,7 +490,49 @@ export default class Marker extends Evented {
+            return this._map._clickTolerance;
+        }
+        return null;
+    }
```
Deleted this method.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -483,7 +490,49 @@ export default class Marker extends Evented {
+        const numeric = +value;
+        if (!Number.isNaN(numeric)) {
+            this._clickTolerance = numeric;
+        }
```
Deleted this method.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -483,7 +490,49 @@ export default class Marker extends Evented {
+
+    /**
+     * Set the `Marker`'s click tolerance.
+     * @param {number} value the click tolerance in pixels
```
Deleted this method.

ChristopherChudzicki commented on src/ui/marker.js:
```diff
@@ -483,7 +490,49 @@ export default class Marker extends Evented {
+        const clickTolerance = this.getClickTolerance();
+        if (clickTolerance === null) return false;
+        return point.distSqr(this._pointerdownPos) >= clickTolerance ** 2;
+    }
```
Deleted this method.

ChristopherChudzicki commented:
@ryanhamley Sorry it took me so long to get back to this. I believe the most recent commit addressed all the comments, except https://github.com/mapbox/mapbox-gl-js/pull/9640#discussion_r438370761 for the reason described there.

**Question for you:** Your [comment](https://github.com/mapbox/mapbox-gl-js/pull/9640#discussion_r438368207) about keeping the API small hit home with me, and now I am wondering whether mapbox benefits much from marker-specific click tolerances at all. To be clear: **I strongly believe marker dragging should respect SOME click tolerance**. But maybe markers should ALWAYS use the map click tolerance, rather than having their own.

Pros:
 - simpler interface
 - probably what people actually want?

Cons:
 - a "more" breaking change, in that the existing behavior of 0-tolerance markers, nonzero-tolerance maps cannot be recovered.

If you think that ALWAYS using the map's click tolerance for markers is a better way to go, I am happy to make a separate, cleaner PR for that and I'll try to do it much sooner than I got around to making these changes.

ChristopherChudzicki commented:
@ryanhamley any thoughts on https://github.com/mapbox/mapbox-gl-js/pull/9640#issuecomment-663892069, or if per-marker tolerance still seems worth while, any chance this might be reviewed again soon?

ryanhamley reviewed (approved):
Hi @ChristopherChudzicki thanks so much for your diligence and patience. I think this looks great. I'm all for keeping the API minimal, but this is just adding one new option to the marker constructor so I think it strikes a nice balance between minimalism and flexibility. The tests look good too.

> I think there should be a separate issue / PR to suppress map click events when dragging markers

If you want to submit a new PR to address this as well, I should have more bandwidth in the future than I did the past couple months and I'll gladly take a look.

Thanks again for the contribution and good work!

ChristopherChudzicki commented:
@ryanhamley Glad this got merged! I'd be happy to submit another PR to supress the map clicks when dragging markers.

nstele commented:
Is there any similar for Android SDK symbolManager?
